camerakit/src/main/api16/com/flurgle/camerakit/Camera1.java

```diff
@@ -419,27 +419,52 @@ private void adjustCameraParameters() {
         initResolutions();
 
         boolean invertPreviewSizes = mDisplayOrientation%180 != 0;
-        mPreview.setTruePreviewSize(
-                invertPreviewSizes? getPreviewResolution().getHeight() : getPreviewResolution().getWidth(),
-                invertPreviewSizes? getPreviewResolution().getWidth() : getPreviewResolution().getHeight()
-        );
-
-        mCameraParameters.setPreviewSize(
-                getPreviewResolution().getWidth(),
-                getPreviewResolution().getHeight()
-        );
-
-        mCameraParameters.setPictureSize(
-                getCaptureResolution().getWidth(),
-                getCaptureResolution().getHeight()
-        );
+        Camera.Parameters resolutionLess = mCamera.getParameters();
+        boolean have_to_readjust = false;
+        if (getPreviewResolution() != null) {
+            mPreview.setTruePreviewSize(
+                    invertPreviewSizes ? getPreviewResolution().getHeight() : getPreviewResolution().getWidth(),
+                    invertPreviewSizes ? getPreviewResolution().getWidth() : getPreviewResolution().getHeight()
+            );
+
+            mCameraParameters.setPreviewSize(
+                    getPreviewResolution().getWidth(),
+                    getPreviewResolution().getHeight()
+            );
+            try{
+                mCamera.setParameters(mCameraParameters);
+            }catch(Exception e){
+                e.printStackTrace();
+                mCameraParameters = resolutionLess; //some phones can't set parameters that camerakit has chosen, so fallback to defaults
+            }
+        }else{
+            have_to_readjust = true;
+        }
+        if (getCaptureResolution() != null) {
+            mCameraParameters.setPictureSize(
+                    getCaptureResolution().getWidth(),
+                    getCaptureResolution().getHeight()
+            );
+            try{
+                mCamera.setParameters(mCameraParameters);
+            }catch(Exception e){
+                e.printStackTrace();
+                mCameraParameters = resolutionLess; //some phones can't set parameters that camerakit has chosen, so fallback to defaults
+            }
+        }else{
+            have_to_readjust = true;
+        }
         int rotation = calculateCaptureRotation();
         mCameraParameters.setRotation(rotation);
 
         setFocus(mFocus);
         setFlash(mFlash);
 
         mCamera.setParameters(mCameraParameters);
+        if (have_to_readjust){
```

Comment:
Could this lead to an infinite loop?

Reply:
Just in theory. In practice camera should have both parameters initialized to just be able to show preview and make pictures, so that is fallback solution if 100ms for syncronisation isn't enough. I could add a counter to remove any possibility.

Reply:
Thanks for the explanation 👍 

> I could add a counter to remove any possibility.

I think that would be ideal, what would be the adequate upper limit of tries?

Reply:
I've added counter, which will take 100 tries 1ms each. This way it won't take too long for user to be bothered with it and it should be enough to be useful.